resolver.go

```diff
@@ -140,38 +142,67 @@ func (r *resolver) Start() error {
 	if r.err != nil {
 		return r.err
 	}
+	// startCh is to seralize resolver Start and Stop
+	r.Lock()
+	r.startCh = make(chan struct{})
```

Comment:
I do not understand the problem in serializing any `Start()/Stop()` sequence which may span across container restart. 

Given the base code does not support concurrent `Start()/Stop()`, the safest approach is in fact serializing this  couple of functions, regardless.

Also, I am not comfortable with the amount of code refactoring in this PR given it targets a panic fix for docker 1.13 while we are in the 1.13 RC process. 

In less abstract terms, I am comparing this changes with the following minimal change: https://github.com/aboch/libnetwork_new/commit/e424dce7938d2cbc5ec003354597c3e98a343939